feat:Add a hardware button support to increase the printlimit

CheatingBoy6890 opened the pull request:
- [X] I have read the [contributing guidelines](https://github.com/PhotoboothProject/photobooth/blob/dev/CONTRIBUTING.md).


- [ ] Documentation update
- [ ] Bug fix
- [X] New feature
- [ ] Other, please explain:


#### What changes did you make? (Give an overview)

Added a new command to the remotebuzzer server (assets/js/remotebuzzer-server.js)
that allows increasing the print limit with a simple GET Request.

#### Is there anything you'd like reviewers to focus on?

The way I interact with 'my.config.inc.php' isn't really good. Maybe you know a better way.

CheatingBoy6890 committed:
Use path module to get Path independent of installation directory

CheatingBoy6890 committed:
Fixed spelling

CheatingBoy6890 committed:
Fixed url handling

andi34 commented:
Hey and thanks for your pull request.
You don't need to increase the limit, it should be fine to unlock print only and it'll respect the defined limit until it's reached again.

Example of an limit of 10 prints, once it's reached 10 print commands executed it'll lock print.
If you're unlocking print now, you'll be able to print again 10x (new limit = 20).

CheatingBoy6890 commented:
My use case is a bit different. I built a small ESP32 powered coin acceptor to increase the print limit when coins are inserted. Different coins give different amounts of pictures you can print.

andi34 commented:
Ah ok, that makes sense. You want to share the ino and used parts to the project? Without it wouldn't make much sense to include the PR

CheatingBoy6890 commented:
Sure here is a github repo I made for this. https://github.com/CheatingBoy6890/Muenzeinwurf

andi34 commented:
New API endpoint to be used
```php
<?php

/** @var array $config */
require_once '../lib/boot.php';

use Photobooth\Environment;
use Photobooth\Service\ConfigurationService;
use Photobooth\Service\LoggerService;
use Photobooth\Service\PrintManagerService;
use Photobooth\Service\ProcessService;
use Photobooth\Utility\PathUtility;

header('Content-Type: application/json');

$loggerService = LoggerService::getInstance();
$configurationService = ConfigurationService::getInstance();
$printManager = PrintManagerService::getInstance();

$logger = $loggerService->getLogger('main');
$logger->debug(basename($_SERVER['PHP_SELF']));

$printCount = $_GET['printCount'] ?? null;

// Validate printCount
if ($printCount === null || !filter_var($printCount, FILTER_VALIDATE_INT, ["options" => ["min_range" => 0]])) {
    $LogData = [
        'error' => 'Invalid printCount. It must be a non-negative integer.',
    ];
    http_response_code(400);
    die(json_encode($LogData));
}

$printCount = (int) $printCount;

$logger->debug('Saving Photobooth configuration with new print limit');
$config['print']['limit'] = ($config['print']['limit'] ?? 0) + $printCount;

try {
    $configurationService->update($config);
    $printManager->unlockPrint();
    $logger->debug('New config saved.');
    echo json_encode([
        'status' => 'success',

        'message' => 'New config saved.',
    ]);
} catch (\Exception $exception) {
    $logger->error('ERROR: Config can not be saved!');
    echo json_encode([
        'status' => 'error',
            'message' => $exception->getMessage(),
    ]);
}

// Kill service daemons after config has changed
ProcessService::getInstance()->shutdown();
exit();
```

CheatingBoy6890 commented:
Thank, I will test this. One concern I have is what happens when multiple coins are inserted before a user decides to start a print.
For example a user inserts 4* 50ct (4* 1 picture in my use case) => The print limit will be set to 1 four times in a row, and in the end the user can only print one picture.
Maybe I also see this wrong, as I'm really not good at php.

andi34 commented:
Maybe 
```
ProcessService::getInstance()->shutdown();
```
Need to be removed. Not tested yet.

andi34 commented:
We can also do some math function to increase the current defined limit. Should be the safest way. Will do later, one line to change

andi34 commented:
```php
$config['print']['limit'] = ($config['print']['limit'] ?? 0) + $printCount;
```
Adjusted

andi34 commented on assets/js/remotebuzzer-server.js:
```diff
@@ -205,7 +205,10 @@ const requestListener = function (req, res) {
     }
 
-    switch (req.url) {
+    const urlObj = new URL(req.url, 'http://localhost/'); //The consctructor requires to input a base url
```
```
const urlObj = new URL(req.url, 'http://' + config.remotebuzzer.serverip + '/');
```

andi34 commented on assets/js/remotebuzzer-server.js:
```diff
@@ -232,6 +236,42 @@ const requestListener = function (req, res) {
+                    }
+                    //Read config file and find the current print limit
+
+                    const configPath = path.join(__dirname, '..', '..', 'config', 'my.config.inc.php');
```
Not needed with the new API endpoint. Only needs a get/post request with `i` to increase the amount.

andi34 committed:
api(print): add endpoint to increase limit

- added new API endpoint to increase the print limit and unlock
- added hidden config to store Photobooth URL

Change-Id: I9cdda9a1aa1f7bb55b21d68ca82cce235be17b3a

andi34 committed:
remotebuzzer: use new api endpoint and config

Change-Id: Ibe1e9145eca3a36657875e588c3c2066444b398a

andi34 commented:
Please try with 
https://github.com/PhotoboothProject/photobooth/commit/323e10a5e88995faaddf060860d6ef772c8e41f8
And
https://github.com/PhotoboothProject/photobooth/commit/5f0396857bbe7d7e859b45a87964158e0d8ece45

Pushed to https://github.com/PhotoboothProject/photobooth/tree/pull/1209

andi34 commented:
Fixup https://github.com/PhotoboothProject/photobooth/commit/083ddd6aa85af9c865e3589a7934aee3a21a7c6b

CheatingBoy6890 commented:
Almost working. When I tested the remote-buzzer-server crashed, because the urlObj constructor also needs a protocol.
I fixed it by using
`const urlObj = new URL(req.url, 'http://' + config.webserver.ip);`
Then it worked fine.

andi34 commented:
Sry, I am wrong

andi34 commented on assets/js/remotebuzzer-server.js:
```diff
@@ -205,7 +205,7 @@ const requestListener = function (req, res) {
     }
 
-    const urlObj = new URL(req.url, 'http://localhost/'); //The consctructor requires to input a base url
+    const urlObj = new URL(req.url, 'http://' + config.webserver.ip);
```
```
config.remotebuzzer.serverip
```
Own IP or from Remotebuzzer server (might be running on a different device and Photobooth only acts as client)

andi34 commented:
Should be considered to adjust as post request.

Mamo800503 commented:
I have trouble with my BL-616 :(
I tried to use BL-616 with ESP32. I connected the BL-616  with gpio4 (white cable). 
When I did, my serial monitor shows me: 
Unknown pulse count: 249
Unknown pulse count: 480
Unknown pulse count: 229
Unknown pulse count: 752
....

CheatingBoy6890 commented:
I think I have a mistake in the wiring diagram. I think the gpio for the reconfigure button and coin acceptor are swapped. So you could try changing PULSE_PIN to 10. Additionally if you don't have an external pull-up try using 
`pinmode(PULSE_PIN, INPUT_PULLUP)`
